### ISSUE TITLE
Custom cache keys or support for multiple schemas in a single database

### ISSUE THREAD

**Issue by owais** [Suor/django-cacheops#120]
Postgresql support multiple schemas. The schemas can have same table names. Two rows from similar tables from 2 different schemas might cause a race condition and return cached object from the WRONG schema.

It can be solved by making it possible to generate the cache key at runtime. We discussed this over at django-cachalot and came up with a solution. May be something similar could be useful done here as well? https://github.com/BertrandBordage/django-cachalot/issues/1

**Comment by tpyo** [comment]
A solution for multiple schemas would be most useful, using numbered redis dbs isn't really a desirable workaround.

**Comment by grjones** [comment]
I also don't know why this issue would be closed.  Using multiple redis databases doesn't seem like a very good or easy solution to implement seamlessly.

**Comment by Suor** [comment]
This issue is closed because there is no official or de facto standard way of using PostgreSQL schemas in Django. So even asking to solve this for cacheops doesn't make sense until it's solved for Django. 

Also, is there anybody here who needs this? It looks for me as theoretical issue for now.

**Comment by grjones** [comment]
Hi @Suor.  Thanks for the quick reply.  First I should say that I love django-cacheops.  I've used it in a couple of projects and it has worked very well for us.  I failed to give you context into my current issue, so here goes.  You're right that Django core does not support schemas, but it can be made to work.  My current project is using django-tenants, so that is really the source of the issue.  To me, the solution @owais provided is good enough to allow me to continue using cacheops.  Maybe using multiple schemas isn't popular enough for this to be a priority, which is fair, but for me it's a +1. Thanks!

**Comment by Suor** [comment]
The fact that you was able to use cachalot says that cacheops is probably overkill for you anyway. 

I'll reopen this and mark it as looking for more interest. We'll see if any more people need/want this.

**Comment by grjones** [comment]
Thanks for reopening @Suor.  Fwiw, I'm not using cachalot.  I just liked their idea making the key configurable.

**Comment by subramanyamVemu** [comment]
Please review if this fix is good 

https://github.com/Suor/django-cacheops/pull/189/commits/ff80570ce0d416558427d2b48d053e46f261e5d7

**Comment by subramanyamVemu** [comment]
Please review this 
https://github.com/Suor/django-cacheops/pull/189/commits/82fd6772afd6f72c0407fa496f58942be6e48595

**Comment by meizon** [comment]
I have a Django project that shares the database for many users and a custom cache key would allow me to prevent invalidation for every user when just one user updates their data.

**Comment by meizon** [comment]
Any more interest here? I'd love to use cacheops, but I don't really want to have to monkey-patch or maintain my own fork.

**Comment by owais** [comment]
> Also, is there anybody here who needs this? It looks for me as theoretical issue for now.

My original use case was very much real and not theoretical at all. Lack of this ability to influence the generation of cache-keys made this otherwise super library useless for me. I've since then moved to a different job and don't need it now but I know the project I left behind will still not be able to use cacheops if they wanted to until this feature is added.

**Comment by meizon** [comment]
Yeah. I have the use case of supporting users who share tables and don't want to invalidate all of the table cache for all users when one user updates their data.

**Comment by Suor** [comment]
@meizon your case should be handled automatically. If you share table then you query it by condition and cacheops does its thing.

**Comment by meizon** [comment]
@Suor Ah, I wasn't sure how cacheops invalidated caches, because when I save a "Resource" how does cacheops know to invalidate that particuar resource for "Resource.objects.for_user(user)" ?

**Comment by Suor** [comment]
@meizin don't know what `.for_user()`, but if that adds condition like `.filter(user=user)` then that will work. Anyway, why you don't just try how it works?

**Comment by meizon** [comment]
Yeah. That's exactly what it does. I'll take a look with some debugging. Thanks!

**Comment by grjones** [comment]
@Suor Maybe I'm misunderstanding @meizon's question, but invalidation with cacheops always happens at the table-level (not row-level), correct?

**Comment by meizon** [comment]
@grjones yes, that's what I assumed.

**Comment by Suor** [comment]
@grjones @meizon invalidation happens on query level based by conditions. It is explained in detail here - http://hackflow.com/blog/2014/03/09/on-orm-cache-invalidation/

**Comment by meizon** [comment]
@suor thanks for that post. Going to go and make sure I understand it all 😃

**Comment by santiagomalter** [comment]
I'm using `django-tenant-schemas` which allow multi-tenant architecture with Django, using Postgres schemas.

Schema-aware database caching would be awesome :)

**Comment by SaidTayebi** [comment]
Would be great to have a Schema-aware database cahing !! The project I'm working on will defenitely need this feature !

**Comment by Suor** [comment]
This has been implemented in 29ad33f635beeebf6a71b55ef564cb87309dd69f, no docs yet. Used like:
```python
CACHEOPS_PREFIX = lambda query: ...
# or
CACHEOPS_PREFIX = 'some.module.prefix_func'
```
`query` contains dbs, tables and conds of a query. Could be completely ignored also, and some threadlocal with request/schema used as prefix:
```python
def prefix_func(query):
    return _request_context.request.get_host()

# Simple middleware to store request
_request_context = threading.local()

class StoreRequestMiddleware:
    def process_request(self, request):
        _request_context.request = request

    def process_response(self, request, response):
        del _request_contexts.request

    def process_exception(self, request, exception):
        del _request_contexts.request
```
If anyone is interested to try this I will be glad to hear about your experience. Also, take a look at [CHANGELOG](https://github.com/Suor/django-cacheops/blob/1b273761aa6490d8633ad763986fe23e1ea91b06/CHANGELOG) it has some backwards incompatible changes.

**Comment by roelvdboom** [comment]
This thread was really helpful! I have implemented the `cacheops_prefix` like this to get it working for `django-tenant-schemas`:

```python
from django.db import connection


def cacheops_prefix(_):
    return connection.tenant.schema_name
```

This way each query will only be cached for its own tenant. I moved the function to a separate file to prevent import errors for `connection` in the `settings.py` file.
